v2/tools/asoctl/internal/crd/cleaner_test.go

```diff
@@ -0,0 +1,311 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package crd
+
+import (
+	"context"
+	"testing"
+
+	resources "github.com/Azure/azure-service-operator/v2/api/resources/v1alpha1api20200601"
+	"github.com/Azure/azure-service-operator/v2/internal/controllers"
+	"github.com/Azure/go-autorest/autorest/to"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/fake"
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	fake2 "sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+// TODO: Currently we need to create clientsets for each test as they run in parallel and we run into `resource already exists` error.
+// TODO: We may require a testing suite re-use the clientsets efficiently.
+func getClientSets() (apiextensions.ApiextensionsV1Interface, client.WithWatch, *Cleaner) {
+	fakeApiExtClient := fake.NewSimpleClientset().ApiextensionsV1()
+	fakeClient := fake2.NewClientBuilder().WithScheme(controllers.CreateScheme()).Build()
+	cleaner := NewCleaner(fakeApiExtClient.CustomResourceDefinitions(), fakeClient, false)
+	return fakeApiExtClient, fakeClient, cleaner
+}
+
+func Test_CleanDeprecatedCRDVersions_CleansAlphaVersion_IfExists(t *testing.T) {
+	t.Parallel()
+
+	fakeApiExtClient, _, cleaner := getClientSets()
+
+	asserter := NewGomegaWithT(t)
+
+	alphaVersion := "v1alpha1api20200601"
+	betaVersion := "v1beta20200601"
+
+	definition := newCRDWithStoredVersions("v1alpha1api20200601", "v1beta20200601")
+
+	_, err := fakeApiExtClient.CustomResourceDefinitions().Create(context.TODO(), definition, metav1.CreateOptions{})
+	asserter.Expect(err).To(BeNil())
+
+	err = cleaner.Run(context.TODO())
+	asserter.Expect(err).To(BeNil())
+
+	crd, err := fakeApiExtClient.CustomResourceDefinitions().Get(context.TODO(), definition.Name, metav1.GetOptions{})
+	asserter.Expect(err).To(BeNil())
+
+	asserter.Expect(crd.Status.StoredVersions).ToNot(BeNil())
+	asserter.Expect(crd.Status.StoredVersions).ToNot(BeEquivalentTo(definition.Status.StoredVersions))
+	asserter.Expect(crd.Status.StoredVersions).ToNot(ContainElement(alphaVersion))
+	asserter.Expect(crd.Status.StoredVersions).To(ContainElement(betaVersion))
+}
+
+func Test_MigrateDeprecatedCRDResources_MigrateAlphaVersion_IfStorage(t *testing.T) {
+	t.Parallel()
+	asserter := NewGomegaWithT(t)
+
+	fakeApiExtClient, fakeClient, cleaner := getClientSets()
+
+	alphaVersion := "v1alpha1api20200601"
+	betaVersion := "v1beta20200601"
+
+	// create CRD
+	definition := newCRDWithStoredVersions(alphaVersion, betaVersion)
+	definition.Spec.Versions = []v1.CustomResourceDefinitionVersion{
```

Comment:
I think you have this test and the next test backwards? AFAIK, `Storage` will only be set for 1 CRD version. We have to make sure that it's _not_ set for the version we're removing, because that's the version that is saved into etcd. If alpha is storage, and we do the migration, then everything in etcd will be alpha. Then we delete alpha and boom, that's bad.

We want the reverse: we're removing alpha, make sure alpha is _not_ storage. Then do migrate and then we're good.